File: ios/CarbBookCore/Sources/CarbBookCore/ImageStack.swift
import Foundation

/// Mirrors packages/core/src/imageStack.ts: which photos an overlapping stack draws, and in what
/// order. Layout, size and colour are the view's business; this decides only the ordering.
public enum ImageStack {
    /// Photos drawn per stack; same value as TS `IMAGE_STACK_MAX`.
    public static let max = 3
}

public struct StackEntry: Sendable {
    public var imageId: String?
    public var carbs: Double?

    public init(imageId: String?, carbs: Double?) {
        self.imageId = imageId
        self.carbs = carbs
    }
}

public struct StackLayout: Sendable, Equatable {
    /// Photos to draw, frontmost first.
    public var imageIds: [String]
    /// How many items in the group those photos do not account for.
    public var overflow: Int

    public init(imageIds: [String], overflow: Int) {
        self.imageIds = imageIds
        self.overflow = overflow
    }
}

/// The item driving the insulin dose is the one worth recognising first, so the biggest carbohydrate
/// contribution goes frontmost. Items with no image still count towards the overflow badge — they
/// exist, they simply have nothing to show (a food without a photo, or a quick-carb row that
/// references no food at all).
public func imageStackLayout(_ entries: [StackEntry], max maxPhotos: Int = ImageStack.max) -> StackLayout {
    // Carry the original index: Swift's sort is not stable, and the stack must not jitter between
    // renders when nothing has changed.
    let withImage = entries.enumerated()
        .filter { _, entry in
            guard let id = entry.imageId else { return false }
            return !id.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        }
        .sorted { a, b in
            // An incomplete food still has an image worth showing, so unknown carbs sort last
            // rather than being hidden.
            let aKnown = (a.element.carbs?.isFinite ?? false)
            let bKnown = (b.element.carbs?.isFinite ?? false)
            if aKnown, bKnown, a.element.carbs != b.element.carbs { return a.element.carbs! > b.element.carbs! }
            if aKnown != bKnown { return aKnown }
            return a.offset < b.offset
        }

    // The id is a content hash used as a URL path, so it is passed through as stored.
    let imageIds = withImage.prefix(Swift.max(0, maxPhotos)).map { $0.element.imageId! }
    return StackLayout(imageIds: imageIds, overflow: Swift.max(0, entries.count - imageIds.count))
}

File: packages/core/src/imageStack.ts
/**
 * Ordering rule for the overlapping photo stacks drawn where several items collapse onto one row
 * (a logged entry, a meal, a plan slot). Layout, size and colour are the clients' business; this
 * module only decides *which* photos and in what order.
 */

/** Photos drawn per stack. Shared so both clients and the CSS agree on the cap. */
export const IMAGE_STACK_MAX = 3;

export interface StackEntry {
  imageId: string | null;
  carbs: number | null;
}

export interface StackLayout {
  /** Photos to draw, frontmost first. */
  imageIds: string[];
  /** How many items in the group those photos do not account for. */
  overflow: number;
}

/**
 * The item driving the insulin dose is the one worth recognising first, so the biggest carbohydrate
 * contribution goes frontmost. Items with no image still count towards the overflow badge — they
 * exist, they simply have nothing to show (a food without a photo, or a quick-carb row that
 * references no food at all).
 */
export function imageStackLayout(entries: StackEntry[], max: number = IMAGE_STACK_MAX): StackLayout {
  const withImage = entries
    // Keep the original index so ties can be broken by it rather than by the sort's whim.
    .map((entry, index) => ({ entry, index }))
    .filter(({ entry }) => typeof entry.imageId === 'string' && entry.imageId.trim() !== '');

  withImage.sort((a, b) => {
    const ca = a.entry.carbs;
    const cb = b.entry.carbs;
    // An incomplete food still has an image worth showing, so unknown carbs sort last rather than
    // being hidden.
    const aKnown = typeof ca === 'number' && Number.isFinite(ca);
    const bKnown = typeof cb === 'number' && Number.isFinite(cb);
    if (aKnown && bKnown && ca !== cb) return cb! - ca!;
    if (aKnown !== bKnown) return aKnown ? -1 : 1;
    // Stable across renders, so the stack does not jitter when nothing has changed.
    return a.index - b.index;
  });

  // The id is a content hash used as a URL path, so it is passed through as stored.
  const imageIds = withImage.slice(0, Math.max(0, max)).map(({ entry }) => entry.imageId!);
  return { imageIds, overflow: Math.max(0, entries.length - imageIds.length) };
}
